shell/platform/windows/compositor_software.cc

```diff
@@ -47,18 +49,101 @@ bool CompositorSoftware::Present(FlutterWindowsView* view,
     return view->ClearSoftwareBitmap();
   }
 
-  // TODO: Support compositing layers and platform views.
-  // See: https://github.com/flutter/flutter/issues/31713
-  FML_DCHECK(layers_count == 1);
-  FML_DCHECK(layers[0]->offset.x == 0 && layers[0]->offset.y == 0);
-  FML_DCHECK(layers[0]->type == kFlutterLayerContentTypeBackingStore);
-  FML_DCHECK(layers[0]->backing_store->type ==
-             kFlutterBackingStoreTypeSoftware);
+  // Bypass composition logic if there is only one layer.
+  if (layers_count == 1) {
+    const FlutterLayer* layer = layers[0];
+    FML_DCHECK(layer != nullptr);
+    if (layer->type == kFlutterLayerContentTypeBackingStore &&
+        layer->offset.x == 0 && layer->offset.y == 0) {
+      auto& backing_store = *layer->backing_store;
+      FML_DCHECK(backing_store.type == kFlutterBackingStoreTypeSoftware);
+      auto& software = backing_store.software;
+      return view->PresentSoftwareBitmap(software.allocation,
+                                         software.row_bytes, software.height);
+    }
+  }
+
+  // Composite many layers.
+  int x_min = INT_MAX;
+  int x_max = INT_MIN;
+  int y_min = INT_MAX;
+  int y_max = INT_MIN;
+  for (const FlutterLayer** layer = layers; layer < layers + layers_count;
+       layer++) {
+    const FlutterPoint& offset = (*layer)->offset;
+    const FlutterSize& size = (*layer)->size;
+    // FlutterPoint and FlutterSize store coordinates as doubles.
+    // Coordinates must be truncated to integers to represent whole pixels.
+    x_min = std::min(x_min, static_cast<int>(offset.x));
+    y_min = std::min(y_min, static_cast<int>(offset.y));
+    x_max = std::max(x_max, static_cast<int>(offset.x + size.width));
+    y_max = std::max(y_max, static_cast<int>(offset.y + size.height));
+  }
+
+  int width = x_max - x_min;
+  int height = y_max - y_min;
+  std::vector<uint32_t> allocation(width * height, kOpaqueBlack);
+
+  for (const FlutterLayer** layer = layers; layer < layers + layers_count;
+       layer++) {
+    // TODO(schectman): handle platform view type layers.
+    // https://github.com/flutter/flutter/issues/143375
+    if ((*layer)->type == kFlutterLayerContentTypeBackingStore) {
+      BlendLayer(allocation, **layer, x_min, y_min, width, height);
+    } else {
+      FML_UNREACHABLE();
+      return false;
+    }
+  }
+
+  return view->PresentSoftwareBitmap(static_cast<void*>(allocation.data()),
+                                     width * 4, height);
+}
+
+void CompositorSoftware::BlendLayer(std::vector<uint32_t>& allocation,
```

Comment:
Instead of an instance method that's declared in the header, should we make this be a top-level function in an anonymous namespace in this file?

In my mind the header file is best for public APIs or methods that need to interact with instance fields. However, this helper is self-contained logic that doesn't require any instance members.